File: myonlinefm/src/main/java/com/sec/myonlinefm/OnLineFMPlayerListener/OberverOnLinePlayerManager.java
package com.sec.myonlinefm.OnLineFMPlayerListener;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by SRC-TJ-MM-BinYang on 2018/2/7.
 */

public class OberverOnLinePlayerManager implements SubPlayerListener {
    private static OberverOnLinePlayerManager observerManager;
    //观察者接口集合
    private List<ObserverPlayerListener> list = new ArrayList<>();

    /**
     * 单例
     */
    public static OberverOnLinePlayerManager getInstance(){
        if (null == observerManager){
            synchronized (OberverOnLinePlayerManager.class){
                if (null == observerManager){
                    observerManager = new OberverOnLinePlayerManager();
                }
            }
        }
        return observerManager;
    }

    @Override
    public void add(ObserverPlayerListener observerListener) {
        list.add(observerListener);
    }

    @Override
    public void notifyObserver(String mPlayer_Url, int play_type) {
        for (ObserverPlayerListener observerListener : list){
            observerListener.observerStartPlayer(mPlayer_Url, play_type);
        }
    }

    @Override
    public void notifyObserverDNS(boolean isStart) {
        for (ObserverPlayerListener observerListener : list){
            observerListener.observerStartDNS(isStart);
        }
    }

    @Override
    public void remove(ObserverPlayerListener observerListener) {
        if (list.contains(observerListener)){
            list.remove(observerListener);
        }
    }
}

File: myonlinefm/src/main/java/com/sec/myonlinefm/OnLineFMPlayerListener/OberverOnLinePlayerManager.kt
package com.sec.myonlinefm.OnLineFMPlayerListener

/**
 * Created by SRC-TJ-MM-BinYang on 2018/2/7.
 */
class OberverOnLinePlayerManager : SubPlayerListener {
    //观察者接口集合
    private val list: MutableList<ObserverPlayerListener?>? = ArrayList()
    override fun add(observerListener: ObserverPlayerListener?) {
        list!!.add(observerListener)
    }

    override fun notifyObserver(mPlayer_Url: String?, play_type: Int) {
        for (observerListener in list!!) {
            observerListener!!.observerStartPlayer(mPlayer_Url, play_type)
        }
    }

    override fun notifyObserverDNS(isStart: Boolean) {
        for (observerListener in list!!) {
            observerListener!!.observerStartDNS(isStart)
        }
    }

    override fun remove(observerListener: ObserverPlayerListener?) {
        if (list!!.contains(observerListener)) {
            list.remove(observerListener)
        }
    }

    companion object {
        private var observerManager: OberverOnLinePlayerManager? = null

        /**
         * 单例
         */
        fun getInstance(): OberverOnLinePlayerManager? {
            if (null == observerManager) {
                synchronized(OberverOnLinePlayerManager::class.java) {
                    if (null == observerManager) {
                        observerManager = OberverOnLinePlayerManager()
                    }
                }
            }
            return observerManager
        }
    }
}